src/libsystemd-network/ndisc-option.c

```diff
@@ -1358,6 +1358,8 @@ static int ndisc_option_parse_encrypted_dns(Set **options, size_t offset, size_t
         r = ndisc_get_dns_name(opt + off, ilen, &res.auth_name);
         if (r < 0)
                 return r;
+        if (!in_charset(res.auth_name, URI_UNRESERVED))
```

Comment:
Not sure, but is it necessary? Are there any RFC that we should refuse such dns name?
If so, how about to call it in dns_name_from_wire_format() ?
Even if it is not prohibited, I think it is OK to refuse dns or domain name with spurious characters, but not sure.
Note, I am planning that most of serialization/deserialization logic in networkd will be moved to JSON format. And, though I am not familiar, but JSON string does not allow some spurious characters. Hence, applying such restriction may be OK even if no RFC prohibit fancy URIs.

Reply:
Also, there was something related comment by @evverx.
https://github.com/systemd/systemd/pull/29495#discussion_r1349851171

Reply:
I switched it to use `dns_name_is_valid_ldh`, which should be better. This part is just a dns name unlike the dohpath parameter which is a full uri. The labels are already escaped, so this will reject those names that have silly characters, and embedded dot.

Reply:
> Not sure, but is it necessary? Are there any RFC that we should refuse such dns name?

Regarding RFCs it's true that domain names can be really weird but to me the issue is that those weird domain names aren't shown properly everywhere. For example networkd still accepts search domains with embedded dots and by looking at just the output of `networkctl status` it's hard to tell why, for example, `resolved` rejects search domains with
```sh
host0: Failed to read search domains for the interface, ignoring: Invalid argument
```
It kind of looks normal:
```sh
# networkctl status host0 | grep Sea
              Search Domains: he.he
```
but it's just a single label with an embedded dot:
```sh
# networkctl status host0 --json pretty | jq .SearchDomains.[].Domain
"he\\.he"
```
(Generally it doesn't seem to make sense to accept stuff `resolved` can reject anyway but if it's accepted I think it should be displayed everywhere in a way that makes it clear that something weird is going on)